docs/USAGE-GUIDE.md

```diff
@@ -607,66 +607,29 @@ In TLS1.3, session ticket messages are sent after the handshake as "post-handsha
 
-    In non-blocking mode, s2n-tls expects the callback to not complete its work. If the callback
-    returns a response code of 0 s2n-tls will return **S2N_FAILURE** with **S2N_ERR_T_BLOCKED**
-    error type and **s2n_blocked_status** set to **S2N_BLOCKED_ON_APPLICATION_INPUT**.
-    The handshake is paused and further calls to **s2n_negotiate** will continue to return the
-    same error until **s2n_client_hello_cb_done** is invoked for the **s2n_connection** to resume
-    the handshake. This allows s2n-tls clients to process client_hello without
-    blocking and then resume the handshake at a later time.
-    If any of the connection properties were changed on the basis of the server_name extension then
-    **s2n_connection_server_name_extension_used** must be invoked before marking the callback done.
+### Client Hello Callback
 
-The callback can return a negative value to make s2n-tls terminate the
-handshake early with a fatal handshake failure alert.
+Users can gain access to the Client Hello during the handshake by setting the callback `s2n_config_set_client_hello_cb()` with an implemented `s2n_client_hello_fn`. A common use-case for this is to switch the associated `s2n_config` based on information in the Client Hello. Note that `s2n_connection_server_name_extension_used()` MUST be invoked before exiting the callback if any of the connection properties were changed on the basis of the Server Name extension. If desired, the callback can return a negative value to make s2n-tls terminate the handshake early with a fatal handshake failure alert.
 
-```c
-int s2n_client_hello_cb_done(struct s2n_connection *conn)
-```
-Marks the non-blocking callback as complete.
-Can be invoked from within the callback when operating in non-blocking mode
-to continue the handshake.
+#### Callback Modes
 
-```c
-int s2n_client_server_name_used(struct s2n_connection *conn)
-```
-Indicates that connection properties were changed on the basis of server_name.
-Triggers a s2n-tls server to send the server_name extension. Must be called
-before s2n-tls finishes processing the ClientHello.
+The callback can be invoked in two modes: **S2N_CLIENT_HELLO_CB_BLOCKING**(default) and **S2N_CLIENT_HELLO_CB_NONBLOCKING**. Use `s2n_config_set_client_hello_cb_mode()` to set the desired mode. Blocking mode will invoke the callback only once before the handshake continues. The handshake is therefore "blocked" until the callback returns a response code. Non-blocking mode will cause the handshake to pause at the Client Hello and return control back to the user. Further calls to `s2n_negotiate()` will return **S2N_FAILURE** with **S2N_ERR_T_BLOCKED** error type and **s2n_blocked_status** set to **S2N_BLOCKED_ON_APPLICATION_INPUT**. Only when the callback signals its work is complete by calling `s2n_client_hello_cb_done()` will the handshake continue. 
```

Comment:
"will invoke the callback only once before the handshake continues". This is technically also what non-blocking mode does. Both modes only invoke the callback once.